R/epidist.R

```diff
@@ -1,242 +1,609 @@
+                    pathogen = NA_character_,
+                    epi_dist,
+                    prob_distribution = NA_character_,
+                    prob_distribution_params = NA_real_,
+                    uncertainty = create_epidist_uncertainty(),
+                    summary_stats = create_epidist_summary_stats(),
+                    citation = create_epidist_citation(),
+                    metadata = create_epidist_metadata(),
+                    method_assessment = create_epidist_method_assessment(),
+                    discretise = FALSE,
+                    truncation = NA_real_,
+                    notes = NULL,
+                    auto_calc_params = TRUE) {
 
-  if (params$distribution == "weibull") {
-    param_vector <- c(shape = params$shape, scale = params$scale)
-    cdf_function <- function(x) {
-      stats::pweibull(x, shape = params$shape, scale = params$scale)
-    }
-    pmf_function <- function(x) {
-      cdf_function(x + 1) - cdf_function(x)
-    }
-    pdf_function <- function(x) {
-      stats::dweibull(x, shape = params$shape, scale = params$scale)
-    }
+  # check input
+  checkmate::assert_character(
+    disease,
+    min.len = 1,
+    any.missing = FALSE,
+    null.ok = FALSE
+  )
+  checkmate::assert_character(pathogen)
+  checkmate::assert_character(epi_dist, len = 1)
+  checkmate::assert_character(
+    prob_distribution,
+    min.chars = 1,
+    min.len = 1,
+    max.len = 2
+  )
+  checkmate::assert_list(uncertainty, names = "unique")
+  checkmate::assert_list(
+    summary_stats,
+    types = c("list", "double", "null"),
+    names = "unique"
+  )
+  checkmate::assert_character(citation, len = 1)
+  checkmate::assert_list(metadata)
+  checkmate::assert_list(method_assessment)
+  checkmate::assert_number(truncation, na.ok = TRUE)
+  checkmate::assert_logical(discretise)
+  checkmate::assert_character(notes, null.ok = TRUE)
+
+  # check whether ci has been provided for each parameter
+  # check whether probability params are named or na
+  stopifnot(
+    "uncertainty must be provided for each parameter" =
+      anyNA(uncertainty, recursive = TRUE) ||
+      length(prob_distribution_params) == length(uncertainty),
+    "probability distribution params must be a named vector or NA" =
+      anyNA(prob_distribution_params, recursive = TRUE) ||
+      !is.null(names(prob_distribution_params))
+  )
+
+  # call epidist constructor
+  epidist <- new_epidist(
+    disease = list(
+      disease = disease,
+      pathogen = pathogen
+    ),
+    epi_dist = epi_dist,
+    prob_dist = prob_distribution,
+    prob_dist_params = prob_distribution_params,
+    uncertainty = uncertainty,
+    summary_stats = summary_stats,
+    citation = citation,
+    metadata = metadata,
+    method_assessment = method_assessment,
+    discretise = discretise,
+    truncation = truncation,
+    notes = notes
+  )
+
+  # call epidist validator
+  validate_epidist(epidist = epidist)
+
+  # return epidist object
+  epidist
+}
```

Comment:
Suggestions for **epidist** methods
1. Print method:
	1. Print pathogen name as well, may be useful when a broad category of diseases (e.g. HPAI, EVD) is caused by multiple related pathogens (or strains).
	2. On pathogen name: possibly need a way to indicate when a pathogen is truly unknown (e.g. a limited 'mystery' outbreak), versus when it is not included in the database. In the first case, there could be different sources for the delay distributions (e.g. from initial field epi study), and for the pathogen identity (e.g. later sequencing study) - unsure how to handle these possibly niche cases. Overall, raises questions about the taxonomic resolution of pathogen names - stick with what's reported in the papers?
	3. Print study name as well; helps users check they have the right study when comparing between outbreaks.
	4. Avoid $\gamma$ symbol for gamma distributions - looks cool but may be unnecessarily confusing to new users (esp. as monospace version looks different from usual gamma symbol). Also suggest avoiding `lN` for lognormal - same reasoning although I appreciate the compact representation.
2. Plot method:
	1. Consider `type = "b"` with a line joining the points, easier to read than points alone.
	2. Consider different plot type (column?) for discretised distributions.
3. Summary statistics: Unclear from the description what this refers to - are these the summary statistics of the data reported in the study, and how do these summary stats relate to the distribution parameters? A working example in `create_epidist_summary_stats()` documentation would be very useful, could also be copied here. Overall suggest explaining this better.
4. Automatic parameter calculation: Suggest moving this argument up to just after the `summary_stats` argument as they are related. Also sugget adding some information as to how the parameters are estimated - is this a fitting step or simply conversion? See below under **epidist_db** that this might be slowing down creation from database - suggest informative message to indicate (potentially slow) fitting if such.

Reply:
In description, could also clarify why user might want to define an epidist object in a pipeline (rather than create database and read in). Perhaps if they've estimated parameters, then want to convert into object to pass into existing analysis steps (e.g. episoap-like templates)?

Reply:
Point 1. ii. is an interesting one and something to be considered going forward. I do not think it is within the scope of this PR and instead have made an issue #88 to work on this further.

Reply:
Point 4 is something I've thought about while developing this code and something that needs addressing. I will not make changes to fix this within this PR, but have opened an issue #89 and will resolve in the near future.

Reply:
Thanks for looking into these issues - I don't have concrete suggestions so best to make new issues for further input (from users?), as you've done.

Reply:
Added pathogen and study to `epidist` print method. Changed distribution shorthand for name. Updated `epidist` plotting method. Updated `epidist` documentation to better explain why a user might want to create an `epidist` object using the constructor instead of the database. All resolved in 39733c7b8eed3909ae807a835b29a16b94c77b3d

Reply:
Get this when running `as_epidist(epiparam()[3, ])`, fitting required for params:

```r
Disease: dengue
Pathogen: dengue virus
Epi Distribution: incubation period
Study: Chan Johansson (2012) <10.1371/journal.pone.0050972>
Distribution: lognormal
Parameters:
  mu: -13.8763957698255
  sigma: 1e-10
```

and for the 11th row, no fitting required;

```r
Disease: influenza
Pathogen: influenza-A-H1N1
Epi Distribution: generation time
Study: Lessler et al. (2009) <10.1056/NEJMoa0906089>
Distribution: weibull
Parameters:
  shape: 2.36
  scale: 3.18
```

Minor quibbles about (1) title and sentence case usage, (2) missing "and" between author names, and (3) printed precision of parameters when estimated from fitting step, but best to push to next PR - may be good to make a new issue?

Reply:
Made issue (#92) to make changes in next PR.